File: R/brick_te.R
# =======================================================================================
# BRICK_TE:
# Simple model to simulate contribution of thermosteric expansion (TE) to global sea-
# level rise. (Grinsted, Moore and Jevrejeva 2010, equations 1/2 that were originally
# applied to global sea-level)
# =======================================================================================
#
#  Requires (input variables):
#  - Tg        global temperature [degC]
#
#  Simulates (output variables):
#  - TE        cumulative sea-level contribution from thermosteric expansion [m]
#              since t0 (i=1)
#              by definition Gs(1) = 0
#
#  Internal variables:
#  - TEeq      Equilibrium TE [m]
#
#  Parameters:
#  - a         sensitivity of TE equilibrium [m/K]
#  - b         equilibrium TE [m] for temperature anomaly Tg = 0
#  - invtau    1/time-scale of exponential decay (e-folding time) [years^-1]
#  - TE_0      initial sea level
#  - tstep     time step
# =======================================================================================

brick_te = function(a     = 0.5,
                    b     = 0.0,
                    invtau= 0.005,
                    TE_0  = 0.0,
                    tstep = 1,
                    Tg) {

  np    <- length(Tg)

  TE    <- rep(NA, np)
  TE[1] <- TE_0

  for(i in 2:np){
#    TEeq  <- a*Tg[i-1] + b
    TE[i] <- TE[i-1]  +  tstep * ( ( a*Tg[i-1] + b - TE[i-1]) * invtau) # sample 1/tau

  }

  return(TE)
}

File: fortran/src/brick_te.f90
!=================================================================================
!  Subroutines to run BRICK-TE:
! Simple model to simulate contribution of thermosteric expansion (TE) to global sea-
! level rise. (Grinsted, Moore and Jevrejeva 2010, equations 1/2 that were originally
! applied to global sea-level)
!=================================================================================
!
! Private parameters/variables 'globally' used within module
!
!   tstep     time step
!
!   a         sensitivity of equilibrium volume Veq [m sle/degC]
!   b         equilibrium volume Veq [m sle] for temperature Tg = 0
!   tau       time-scale of exponential decay (e-folding time) [years]
!   TE_0      initial thermosteric expansion
!
!   TE         current thermosteric expansion
!=================================================================================

module brick_te

    USE global
    implicit none
    private
    
! parameters:
    real(DP) :: tstep
    real(DP) :: a, b, tau, TE_0

! variables
    real(DP) :: TE       ! thermosteric expansion
        
! public subroutines
    public :: brick_te_step, init_brick_te


contains


!------------------------------------------------------------------------------
subroutine init_brick_te(time_step, equil_sensitivity, equil_T0, &
                       timescale, Initial_TE, thermosteric)
!  =========================================================================
! |  Initialize the BRICK-TE parameters and initial variables.                                   |
!  =========================================================================

    real(DP), intent(IN) :: time_step
    real(DP), intent(IN) :: equil_sensitivity
    real(DP), intent(IN) :: equil_T0
    real(DP), intent(IN) :: timescale
    real(DP), intent(IN) :: Initial_TE
    
    real(DP), intent(OUT) :: thermosteric
    
  
! Assign values to model parameters
    tstep = time_step
    a     = equil_sensitivity
    b     = equil_T0
    tau   = timescale
    TE_0  = Initial_TE

! Initial values
    thermosteric = TE_0
    TE           = thermosteric
    
end subroutine init_brick_te
!------------------------------------------------------------------------------


!------------------------------------------------------------------------------
subroutine brick_te_step(Tg, thermosteric)
!  ==========================================================================
! | Calculate current state from previous state
! |
! | Input:
! |       Tg:     Greenland/Global mean surface temperature (degC)
! |
! | Output:
! |       thermosteric:    thermosteric expansion [m]
!  ==========================================================================

    implicit none

    real(DP), intent(IN)  :: Tg

    real(DP), intent(OUT) :: thermosteric
    
    real(DP) :: TEeq
    
! Start model
    TEeq  = a * Tg + b                   ! equilibrium TE

    thermosteric = TE + tstep * ((TEeq - TE) / tau)
    TE           = thermosteric


end subroutine brick_te_step
!------------------------------------------------------------------------------

END MODULE brick_te